msal4j-sdk/src/main/java/com/microsoft/aad/msal4j/AadInstanceDiscoveryProvider.java

```diff
@@ -62,40 +62,39 @@ static InstanceDiscoveryMetadataEntry getMetadataEntry(URL authorityUrl,
                                                            boolean validateAuthority,
                                                            MsalRequest msalRequest,
                                                            ServiceBundle serviceBundle) {
-        String host = authorityUrl.getHost();
-
-        if (shouldUseRegionalEndpoint(msalRequest)) {
-            //Server side telemetry requires the result from region discovery when any part of the region API is used
-            String detectedRegion = discoverRegion(msalRequest, serviceBundle);
 
-            if (msalRequest.application().azureRegion() != null) {
-                host = getRegionalizedHost(authorityUrl.getHost(), msalRequest.application().azureRegion());
-            }
+        String host = authorityUrl.getHost();
 
-            //If region autodetection is enabled and a specific region not already set,
-            // set the application's region to the discovered region so that future requests can skip the IMDS endpoint call
-            if (null == msalRequest.application().azureRegion() && msalRequest.application().autoDetectRegion()
-                    && null != detectedRegion) {
-                msalRequest.application().azureRegion = detectedRegion;
-            }
-            cacheRegionInstanceMetadata(authorityUrl.getHost(), msalRequest.application().azureRegion());
-            serviceBundle.getServerSideTelemetry().getCurrentRequest().regionOutcome(
-                    determineRegionOutcome(detectedRegion, msalRequest.application().azureRegion(), msalRequest.application().autoDetectRegion()));
+        //If instanceDiscovery flag set to false or instance discovery previously failed, do not do instance discovery
+        if (!msalRequest.application().instanceDiscovery() || instanceDiscoveryFailed) {
+            return InstanceDiscoveryMetadataEntry.builder().
+                    preferredCache(host).
+                    preferredNetwork(host).
+                    aliases(Collections.singleton(host)).
+                    build();
         }
 
-        InstanceDiscoveryMetadataEntry result = cache.get(host);
+        //If there is no cached instance metadata, do instance and region discovery and cache the result
+        if (cache.get(host) == null) {
+            if (shouldUseRegionalEndpoint(msalRequest)) {
+                //Server side telemetry requires the result from region discovery when any part of the region API is used
+                String detectedRegion = discoverRegion(msalRequest, serviceBundle);
```

Comment:
if the request provides an `azureRegion`, do you still need to do the discovery? Might be better to use the `if(masalRequest.application().azureRegion() != null)` logic before making this call.

```suggestion
            if (msalRequest.application().azureRegion() != null) {
                    host = getRegionalizedHost(authorityUrl.getHost(), msalRequest.application().azureRegion());
            }
            else if (msalRequest.application().autoDetectRegion() ) {
            //Server side telemetry requires the result from region discovery when any part of the region API is used
                String detectedRegion = discoverRegion(msalRequest, serviceBundle);
                msalRequest.application().azureRegion = detectedRegion;
                host = getRegionalizedHost(authorityUrl.getHost(), msalRequest.application().azureRegion());
                }

```

Reply:
I'm having trouble finding the design doc, but there's a list of region telemetry values for different scenarios, including:
- A developer set the region, and MSAL is able to detect the same one
- A developer set the region, but MSAL found a different one
- A developer set the region, but MSAL's autodetect failed for some reason

This means that we need to do autodetection whenever any part of the region API is used, unless we decide to just not collect this telemetry. I believe this behavior is the same in .NET and other MSALs too (feel free to correct me on that @bgavrilMS).

Our RegionTelemetry enum gives a good overview of all the different scenarios we look for: https://github.com/AzureAD/microsoft-authentication-library-for-java/blob/dev/msal4j-sdk/src/main/java/com/microsoft/aad/msal4j/RegionTelemetry.java

Reply:
Yeah, we will try to auto-detect the region even if the app developer tells us "use region X". Auto-detection should only happen **once** per process (even if it fails) and I believe is set to timeout after 2 sec. 

If this is too slow, we can discuss about removing this logic.

Reply:
Prior to these changes we were incorrectly doing autodetection every time a region was used, since the code to do it was happening before the code to check if instance metadata was cached.

With these changes autodetection happens only if no instance is cached, so it'll only happen once for a given authority/region and thus the 2 second timeout will also only happen once.